### ISSUE TITLE
Enable all classes to be autoloaded

### ISSUE THREAD

**Issue by jonasraoni** [pkp/pkp-lib#6091]
**Describe the problem you would like to solve**
Not knowing which classes/constants exist at a given scope, or worse, having to import all of them manually is a possible cause of problems.
If a call located at the beginning of the call stack stops loading a given class/constant, it will create a breaking change, as the deeper calls may not be loading everything they need.

**Describe the solution you'd like**
I would like to have a generic auto_loader, which is able to discover all the existing classes.
To achieve this, I guess all the classes should follow a strict namespace pattern (obeying a naming/folder structure).

TODO:
- [x] Permit autoloading of everything in `classes`
- [x] Permit autoloading of controllers
- [x] Permit autoloading of pages
- [x] Permit autoloading of plugins
- [x] Rename all `.inc.php` to `.php`
  - [x] Controllers
  - [x] Classes
  - [x] Included plugins
  - [x] Major external plugins
  - [x] Pages
  - [x] API Handlers
  - [x] Others
- [x] Scan for and fix unnecessary calls to import
- [x] Survey other potential areas for autoloading
- [x] Fix scanning for mailable class names (see https://github.com/pkp/pkp-lib/pull/7444#discussion_r747413300)

**Comment by asmecher** [comment]
There is already a custom [autoloader in `lib/pkp/include/functions.inc.php`](https://github.com/pkp/pkp-lib/blob/3_2_1-0/includes/functions.inc.php#L321) called `customAutoload`, and it's [installed in `lib/pkp/classes/core/PKPApplication.inc.php`](https://github.com/pkp/pkp-lib/blob/3_2_1-0/classes/core/PKPApplication.inc.php#L203..L212).

It's not used everywhere, though; there are lots of places where `import` statements could be replaced with full namespaces to promote autoloading. Here's [one example](https://github.com/pkp/ojs/blob/3_2_1-1/pages/workflow/WorkflowHandler.inc.php#L73) where the autoloader is used instead of an `import` statement. This will be a matter of ongoing cleanup (and hands are welcome).

I believe there will be a few cases where an autoloader by itself won't be enough, e.g. when unserializing an object when the class definition is not loaded. That's used in some of the `_settings` tables, and will probably be necessary in some of the caching code as well when we reimplement the object caching as part of https://github.com/pkp/pkp-lib/issues/3304 (not yet merged). It's probably the case that when the class specifies a namespace, it's included in the serialized form...

**Comment by jonasraoni** [comment]
Uhum, I've seen this autoloader =]

The `unserialize` is a good call. Out of curiosity, I've just checked the behavior and I think it will be ok (the `serialize` stored the namespace and the `unserialize` called the auto loader).

On my side, I'm trying to add types/namespaces in everything that I'm writing, I understand it's tough... I'm just leaving ideas/suggestions as they come up in my mind before they go to into oblivion.

**Comment by asmecher** [comment]
Here's a proposal, similar to #6092.

Draft PRs:
- https://github.com/pkp/pkp-lib/pull/6853/files
- https://github.com/pkp/ojs/pull/3069/files

Proposal: Add namespaces to core classes in a way that's compatible with the [existing custom autoloader](https://github.com/pkp/pkp-lib/blob/3_2_1-0/includes/functions.inc.php#L321). When `strict` mode (in `config.inc.php`) is not enabled, [use `class_alias`](https://github.com/pkp/pkp-lib/pull/6853/files#diff-117e9322bc0243206d5619fc4ecfb3eb254881327628164b951156e65c139fc5R476) to maintain compatibility.

To introduce this for a given piece of code:
- [Declare a namespace](https://github.com/pkp/pkp-lib/pull/6853/files#diff-117e9322bc0243206d5619fc4ecfb3eb254881327628164b951156e65c139fc5R17) in the class file corresponding to its path. For example, for a class in pkp-lib: `namespace \PKP\somePackage;`
- For compatibility, add a `class_alias` to the bottom of the class to be invoked when not in strict mode.
  ```php
  if (!PKP_STRICT_MODE) {
    class_alias('\PKP\somePackage\SomeClass', '\SomeClass');
  }
- In the core codebase, change references to `SomeClass` so they specify the namespace. (This should not be necessary unless `strict` mode is enabled in `config.inc.php`.)
  ```diff
  -class Issue extends DataObject {
  +class Issue extends \PKP\core\DataObject {
- Remove `import` statements referencing the class so that the autoloader is used instead.
  ```diff
  -import('lib.pkp.classes.core.DataObject');
  ```

**Comment by asmecher** [comment]
@NateWr and @jonasraoni in particular, please have a look at the above proposal!

**Comment by jonasraoni** [comment]
I've read the documentation some times and I can't remember about this `class_alias` function 😂

Looks like it's a safe idea! The only issue that comes to my mind would be something involving the full class name (e.g. serializing/deserializing).

About the autoloader, as we're using composer, perhaps we could use its built-in autoloader and follow a consistent path/naming convention later.

**Comment by NateWr** [comment]
Oh that looks nice! Does the `class_alias` work with extends? For example, right now we've got:

```
class Announcement extends \PKP\core\DataObject {
```

What if a class, in a plugin, did this:

```
class Announcement extends DataObject {
```

Would that still autoload? Or presumably they'll have the `import()` statement still in there? Will that still work for them? (If not, I don't think that's a deal-breaker, but we'll need to carefully document the change and try to put all the migration pain into one release.)

**Comment by asmecher** [comment]
> Does the `class_alias` work with extends?

Yes -- once the `class_alias` is in place, code referring to `DataObject` seamlessly behaves well. However, because the `class_alias` is introduced in the same place as the aliased class is defined, autoloading has a bit of a chicken-and-egg problem to get the `class_alias` loaded in the first place.

Legacy code *should* already contain import statements, but in many cases it'll just assume core classes are already in place (e.g. `DataObject`). Those assumptions will still hold, but via the autoloader mechanism rather than explicit `import`s in core, and where imports are still present in legacy code, it'll all continue to work. There may be a few rare cases we remove `import` statements in core that were legitimately unnecessary there, but were depended on by legacy code e.g. plugins. These cases will result in a missing class error, but will be vanishingly rare, I think.

> What if a class, in a plugin, did this:
> ```
> class Announcement extends DataObject {
> ```
> Would that still autoload?

If the legacy code didn't `import('lib.pkp.classes.core.DataObject)` but instead relied on core to do that, then the assumption still holds, because the same code in core that previously contained the import will still load the class (and introduce the alias) prior to this line of legacy code, but now relying on autoloading.

If the legacy code did `import(...)`, then we're good, because the `import` will still work.

> The only issue that comes to my mind would be something involving the full class name (e.g. serializing/deserializing).

This remains to be proven, but my expectation is that the alias is applied as a mapping when resolving into bytecode, and after that the fully qualified name will be used. But we're not serializing DataObjects much if at all, so I think it's an abstract problem for now. (The [caching changes](https://github.com/pkp/pkp-lib/pull/5743) around object caching support will reveal this when we get back to them, but I don't think we're making any showstopping mistakes here.)

> About the autoloader, as we're using composer, perhaps we could use its built-in autoloader and follow a consistent path/naming convention later.

I'm interested in using the Composer autoloader but don't know much about applying it outside of Composer. We've had a few brief conversations about adopting Composer instead of submodules, but no decisions yet.

We currently have 3 forms of class loading:
- `import`, which we're trying to remove with this approach
- The PKP autoloader, which is a simple mapping of fully qualified classname into path; that's what these changes rely on, but it's already been in place in the codebase for a while
- The Composer autoloader, which uses a filesystem-backed cache of classnames.

I have one worry about adopting Composer more broadly, which is that it doesn't seem to support collecting dependencies from several descriptors at once -- e.g. `pkp-lib`, the application, and plugins -- and that it requires a cache-building step that may not play well with plugins. But I could stand to learn more about this and I'm open to ideas here.

**Comment by NateWr** [comment]
> But we're not serializing DataObjects much if at all

One new area of the codebase that serializes classes is the task queue. Tasks, including the dependencies injected into them, are serialized into the the database. Should be pretty easy to run a test with an announcement that sent the emails in tasks.

**Comment by jonasraoni** [comment]
> We've had a few brief conversations about adopting Composer instead of submodules

Yeah, it's needed to make a list of pros/cons about packing everything as a package, possibly even the plugins could use the same system 🤔
- Simplify the end-user installation (not needed to issue a `submodule update` + `composer install` for specific folders).
- Versioning: Updates could grab the latest non-breaking change automatically.
- More complex to manage customizations (check how composer deals with modified files at the vendor folder)
- [...]

> doesn't seem to support collecting dependencies from several descriptors at once

If we end up converting the modules to packages, it works to have a package for the `pkp-lib` using the namespace `PKP\Xyz` and another package (or the project itself) using the same `PKP\Xyz` namespace.

I just run a small test, using the same namespace and class name to see how it deals with conflicts, and composer gave preference to the class in the package instead of my project.

**Comment by asmecher** [comment]
OPS after mods:
real	15m8.171s
user	30m59.733s
sys	2m54.593s

OPS before mods:

**Comment by asmecher** [comment]
`lib/pkp/classes/core`:
- [ ] announcement
- [ ] cache
- [ ] citation
- [ ] cliTool
- [ ] components
- [ ] config
- [ ] context
- [ ] controlledVocab
- [ ] controllers
- [ ] core
- [ ] db
- [ ] file
- [ ] filter
- [ ] form
- [ ] handler
- [ ] i18n
- [ ] identity
- [ ] install
- [ ] linkAction
- [ ] log
- [ ] mail
- [ ] metadata
- [ ] migration
- [ ] navigationMenu
- [ ] note
- [ ] notification
- [ ] oai
- [ ] payment
- [ ] plugins
- [ ] publication
- [ ] query
- [ ] reviewForm
- [ ] scheduledTask
- [ ] search
- [ ] security
- [ ] services
- [ ] session
- [ ] site
- [ ] stageAssignment
- [ ] statistics
- [ ] submission
- [ ] task
- [ ] template
- [ ] tombstone
- [ ] user
- [ ] validation
- [ ] views
- [ ] webservice
- [ ] workflow
- [ ] xml
- [ ] xslt

**Comment by asmecher** [comment]
Here's a proof of concept PR for the introduction of PSR-4 autoloading: https://github.com/pkp/pkp-lib/pull/7540/files
- It introduces PSR4-based autoloading using the [`composer.json` configuration file](https://github.com/pkp/pkp-lib/pull/7540/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34). The proof of concept only touches classes in `lib/pkp/classes`, but a full implementation would include other areas (`APP\`, controllers, pages).
- It *does not* remove [old autoloader](https://github.com/pkp/pkp-lib/blob/stable-3_3_0/classes/core/PKPApplication.inc.php#L172) (link is to `stable-3_3_0` branch), as this would require all `.inc.php` files in `lib/pkp/classes` to be renamed to `.php` instead. Both can coexist, though I want to remove the old one before 3.4.0 is released.
- It looks like the PHP autoformatter is introducing some new conventions; I might want to disable this while I'm doing batch renaming to avoid rebase problems.

I would like to introduce this to `main` progressively, i.e. have both autoloaders coexist in the next few weeks while I make the full changes. I think this would be less disruptive to others working on their own branches.

@NateWr, does this approach sound OK? Any objections to starting now?

**Comment by NateWr** [comment]
The approach looks good. :+1: 

If possible, it'd be great to wait for a couple of big PRs from @defstat and @ewhanson. Dimitris (#7376) is days away from merge. @ewhanson (#6466) may dip into next year. That said, his work is mostly on new files or the crossref plugin, so may be less impacted.

**Comment by asmecher** [comment]
Thanks, @NateWr. How about I continue working on that PR, including getting it merge-ready, but keep all the actual renames out of it until the other PRs are in? That way once all of us get our merges done it'll be a moment's work to batch-rename everything.

**Comment by NateWr** [comment]
Sounds good if it doesn't slow you down too much.

**Comment by asmecher** [comment]
@NateWr and @Vitaliy-1, could you have a look at this commit?

https://github.com/pkp/pkp-lib/pull/7540/commits/7fe2d6c4b67e1f6048780df678fdd8918a6b3e0c

This is all that's required for us to start renaming `.inc.php` files to `.php` and relying on Composer's PSR-4 autoloader in place of our [custom autoloader](https://github.com/pkp/pkp-lib/blob/stable-3_3_0/classes/core/PKPApplication.inc.php#L172). Both can coexist happily. A few notes:

- There is the potential for collisions, i.e. `\PKP\announcement\xyz` could refer to `lib/pkp/classes/announcement/xyz.php` or `lib/pkp/pages/announcement/xyz.php`. The PSR-4 loader looks for them in sequence and stops when it finds a first match. I think we have 3 options:
  - Rename all our current namespaces from e.g. `\PKP\announcement` to `\PKP\classes\announcement`, which would make every namespace unambiguous. I think that's too inconvenient for coders and would also be a massive `sed` exercise.
  - Or we could use another namespace that doesn't match a directory name for `pages`, `components`, and `plugins`, such as `\PKPPages`, but it feels arbitrary to be mixing conventions like this.
  - Or we could accept the potential for collisions and leave it as is. I favour this option; I don't think our codebase will contain any collisions.
- The other commits [in the PR](https://github.com/pkp/pkp-lib/pull/7540/commits) provide examples for how the codebase will need to change to make use of autoloading. It's pretty minimal.
- Regarding what we include in the 3.4 release for backwards compatibility:
  - We should leave the `import` function in place for 3.4.0, but deprecate it (and make it throw an exception when strict mode is enabled). Lots of plugins etc. will still use this.
  - I think we can remove the old custom autoloader for 3.4.0. There probably isn't much 3rd-party code with namespaces following our patterns in the wild and we've never had a pattern for namespaces in plugins.

**Comment by NateWr** [comment]
Looks good to me. I agree we're not likely to have collisions. And this sets us up well to eventually move handlers into `/classes` (eg - `/classes/pages/X` would have the same namespace).

I left one comment about `HANDLER_CLASS` at https://github.com/pkp/pkp-lib/pull/7540/files#r767806998. But I'm not sure how easy it would be to adjust that in the routing code.

**Comment by Vitaliy-1** [comment]
PR looks good to me either

> There is the potential for collisions, i.e. \PKP\announcement\xyz could refer to lib/pkp/classes/announcement/xyz.php or lib/pkp/pages/announcement/xyz.php

Can you explain this example in more detail: why after these changes it can refer to the latter class? `pages` are prefixed: `"PKP\\pages\\": "pages/"`, so it's `\PKP\pages\announcement\xyz`. Do I understand this right that you mean If there were `pages/announcement` directory in `classes`?

**Comment by asmecher** [comment]
Sorry, @Vitaliy-1, you're right -- my example doesn't work, but yours does. The collision would be possible with namespaces that match directory names within `classes`, e.g. `classes/pages/*.php`, `classes/controllers/*`, `classes/plugins/*`. But I think collisions in these are unlikely enough that we can forget it.

**Comment by asmecher** [comment]
Renaming classes from `.inc.php` to `.php`:
```
for name in `find . -name \*.inc.php | sed -e "s/\.inc.php//"`; do git mv $name.inc.php $name.php; done
```
Looking for missing `use` statements by class name:
```
vi `find lib/pkp/controllers controllers -name \*.php | xargs grep -l -e "[^P\\]Services::" | xargs fgrep -L "use APP\core\Services;"`
```

**Comment by asmecher** [comment]
PRs for renaming pages to .php:
- https://github.com/pkp/ojs/pull/3475
- https://github.com/pkp/pkp-lib/pull/8124

**Comment by asmecher** [comment]
PRs for filter cleanup:
- https://github.com/pkp/ojs/pull/3499
- https://github.com/pkp/pkp-lib/issues/6091

**Comment by asmecher** [comment]
There will be some residual mop-up, but substantively this is completed, and the rest can be done as it's encountered. The main repos don't include any .inc.php files (besides configuration) anymore. :+1:

**Comment by jonasraoni** [comment]
Great 😁

**Comment by steinmb** [comment]
Impressive work! That is one large brave merge :)

**Comment by asmecher** [comment]
@steinmb, on the contrary -- it is several hundred cowardly merges, a lot of failed Travis tests, and ongoing paranoia about new comments on the issue now that it's been closed!

**Comment by bozana** [comment]
@asmecher, we changed the namespacing in the DB table scheduled_tasks, but what about scheduledTasks.xml and thus AcronPlugin? Still to come (with another issue)?

**Comment by asmecher** [comment]
@bozana, I think those are already done -- I had missed the DOAJInfoSender but the others have all been converted. Did I miss something?

**Comment by jonasraoni** [comment]
I've seen some cases in the build log yesterday, now looks like only OPS have them.

**Comment by asmecher** [comment]
Ah, got them!

**Comment by dkikan** [comment]
> Here's a proposal, similar to #6092.
> 
> Draft PRs:
> 
> * https://github.com/pkp/pkp-lib/pull/6853/files
> * https://github.com/pkp/ojs/pull/3069/files
> 
> Proposal: Add namespaces to core classes in a way that's compatible with the [existing custom autoloader](https://github.com/pkp/pkp-lib/blob/3_2_1-0/includes/functions.inc.php#L321). When `strict` mode (in `config.inc.php`) is not enabled, [use `class_alias`](https://github.com/pkp/pkp-lib/pull/6853/files#diff-117e9322bc0243206d5619fc4ecfb3eb254881327628164b951156e65c139fc5R476) to maintain compatibility.
> 
> To introduce this for a given piece of code:
> 
> * [Declare a namespace](https://github.com/pkp/pkp-lib/pull/6853/files#diff-117e9322bc0243206d5619fc4ecfb3eb254881327628164b951156e65c139fc5R17) in the class file corresponding to its path. For example, for a class in pkp-lib: `namespace \PKP\somePackage;`
> * For compatibility, add a `class_alias` to the bottom of the class to be invoked when not in strict mode.
>   ```
>   if (!PKP_STRICT_MODE) {
>       class_alias('\PKP\somePackage\SomeClass', '\SomeClass');
>   }
>   ```
> * In the core codebase, change references to `SomeClass` so they specify the namespace. (This should not be necessary unless `strict` mode is enabled in `config.inc.php`.)
>   ```diff
>   -class Issue extends DataObject {
>   +class Issue extends \PKP\core\DataObject {
>   ```
> * Remove `import` statements referencing the class so that the autoloader is used instead.
>   ```diff
>   -import('lib.pkp.classes.core.DataObject');
>   ```
> 
> Because the `setConstants` function that passes constants along for use in JS/Vue environments expects global constants, it needs to be adapted for constants in namespaces too. To accommodate this I've [adapted it](https://github.com/pkp/pkp-lib/commit/fe5352f445fcb44e5bbbd1ae16de65023dd185d1#diff-8100c829b3c8e2b0f5bbf2d00cbc29f5b67dfa0b7183132bfa8694c7d6e476a7L579) to accept a mapping of CONSTANT_NAME => 'constantValue' rather than a list of constant names alone. Code using `setConstants` will need to be adapted in 3rd-party code (heads-up for the release notebook, @NateWr).
> 
> Additional details on page handler classes:
> 
> Starting with 3.4.0, page handler classes can add autoloading and namespaces. This is optional. Steps to adding namespacing to a page handler ([ojs](https://github.com/pkp/ojs/pull/3285) [pkp-lib](https://github.com/pkp/pkp-lib/pull/7641)):
> 
> 1. Introduce the namespace in the handler class.
>    ![image](https://user-images.githubusercontent.com/200411/150588188-b286b876-0341-4212-89fe-3d5ea172451a.png)
>    Now that the class is in a namespace, you'll have top make sure any required classes are specified in `use` statements.
>    The application pages should be in `APP\pages\...`, and for pkp-lib they should be in `PKP\pages\...`.
> 2. Rename the `.inc.php` handler class to `.php` instead. The PSR-4 autoloader only works on `.php` files, not `.inc.php`.
> 3. Change the `index.php` wrapper to include the namespace:
>    
>    * Where the operation is matched to the handler class, use the fully qualified class name instead (e.g. `\PKP\pages\about\AboutContextHandler` instead of just `AboutContextHandler`)
>    * An optional but nice-to-have clean-up: Rather than defining `HANDLER_CLASS`, you can instantiate and return the handler from inside the `index.php` wrapper:
>      ```
>      return new \PKP\pages\about\AboutContextHandler();`
>      ```
>      
>      
>          
>            
>          
>      
>            
>          
>      
>          
>        
>      (If the handler class's constructor takes a `$request` parameter, the `$request` object should be in scope in the `index.php` wrapper.)

Thanks. I am getting error Warning: constant(): Couldn't find constant submissionPayments in /home/zitco/public_html/interjournals.org/ojs/lib/pkp/classes/template/PKPTemplateManager.inc.php on line 541 in the setConstants function as the constant is not available. Can this function please include some error handling? thanks again. I am using OJS 3.3.0.8

**Comment by jonasraoni** [comment]
@dkikan, could you please post your question on the forum (https://forum.pkp.sfu.ca)? There you'll get better support, and before asking for fixes, you might try to use a newer version.

**Comment by jonasraoni** [comment]
@dkikan, this has been fixed already, just upgrade your installation.

**Comment by dkikan** [comment]
> @dkikan, this has been fixed already, just upgrade your installation.

Thanks for the update. I will upgrade.

**Comment by jonasraoni** [comment]
@asmecher
- Added missing namespaces everywhere
- PHPDoc typehints (only the ones that my editor complained, so it's not covering all)
- The only syntax change was probably replacing the deprecated `"${var}"` by `"{$var}"`.

**PRs for `main`**
- https://github.com/pkp/ojs/pull/3830
- https://github.com/pkp/omp/pull/1365
- https://github.com/pkp/ops/pull/491
- https://github.com/pkp/pkp-lib/pull/8792

**Comment by asmecher** [comment]
@jonasraoni, those changesets are too big to review manually for 3.4, but if you're confident that they're just syntactical, please go ahead.

**Comment by jonasraoni** [comment]
Ok! These PRs are generally safe and they will fix some errors that nobody flagged yet (e.g.```<?php namespace X; throw new Exception('fail');```).

After these ones, I'll open more PRs covering random errors.
